### ISSUE TITLE
Use promises more consistently and thoroughly

### ISSUE THREAD

**Issue by brianhelba** [girder/girder#2030]
This is a proposal to standardize around jQuery 3 Promises as the preferred way to handle asynchronous programming in Girder's web code.

At the moment, there are several issues:
* Girder contains mostly jQuery 2 Promises, with a few ES 6 Promises.
* Many asynchronous methods do not return Promises.
* Even where promises are available, many places use transient event binding instead.
* The jQuery promise API is used inconsistently and with deprecated methods.

To resolve this, it is proposed that we:
* [ ] Eliminate any ES 6 Promises
  * They are not [compatible with IE 11](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Browser_compatibility), which comprises [about 14% of desktop browsers in the US](http://gs.statcounter.com/browser-market-share/desktop/united-states-of-america) and [8.5% in Europe](http://gs.statcounter.com/browser-market-share/desktop/europe). More importantly, some of Girder's users in institutional environments (e.g. hospitals) are forced to use IE. While Girder doesn't explicitly support IE, we don't have to go out of our way to break it either.
  * Their performance (under V8) is currently not superior to third-party promise implementations, and [is often significantly worse](https://softwareengineering.stackexchange.com/a/279003).
  * Most of Girder, and web code in general, uses jQuery for Promises, so there would be a large overhead to consistently switch to ES6 Promises.
  * The alternative [Bluebird promise library](http://bluebirdjs.com/docs/features.html) is probably overkill for now.
    * Its killer `.promisifyAll` feature is directly intended for Node.js.
    * Its map / reduce functions are also probably most useful in a server environment, where lots of similar asynchronous calls may be made in parallel.
    * Its error catching and stack saving features are less important for a front-end app, where uncaught errors are usually not as devastating (though perhaps somewhat important to log).
    * Its `.delay` function is effectively a nice promisified wrapper around `window.setTimeout`. Perhaps we should replicate this as a Girder utility method.

* [ ] Eliminate and stop using `.error` to handle jqXHR Promises.
  * This method has been deprecated for 6 years, [since jQuery 1.8](https://bugs.jquery.com/ticket/9399), and is removed in jQuery 3.
  * `.fail` should be used instead
  * [ ] Since Girder still uses `.error` so prevalently, we should patch it to print a warning when used (which will also allow it to work under jQuery 3).

* [ ] Upgrade to jQuery 3
  * jQuery 3 is finally able to adhere to the [Promises/A+ standard](https://promisesaplus.com/).
  * The updated behavior of `.then` in jQuery 3 (where `throw` becomes a rejection value) is necessary for sane chaining of custom asynchronous stages.

* [ ] Use the jQuery 3 Promise API correctly.
  * Developers should, by default, use [`.done`](http://api.jquery.com/deferred.done/), [`.fail`](http://api.jquery.com/deferred.fail/), and [`.always`](http://api.jquery.com/deferred.always/) to attach handlers to Promises.
  * [`.then`](http://api.jquery.com/deferred.then/) should be used only when it's explicitly necessary: as an intermediate stage in a Promise pipeline, to yield a brand-new Promise.
    * If `.then` returns an ordinary value, it will become the fulfillment value of a new resolved Promise.
    * If `.then` `throw`s a value, it will become the rejection value of a new resolved (rejected) Promise.
    * If `.then` returns a Promise (also known as a "thenable" value), it will become the newly yielded Promise.
    * Do not use `.then` without returning a value. If you want to strip the fulfillment / rejection value off of a Promise while maintaining its state, then explicitly `return undefined` or `throw undefined`.
  * Use the second argument of `.then` (or [`.catch`](http://api.jquery.com/deferred.catch/)) only when explicitly necessary: in the same way as `.then`, but when a previous rejection should be handled and transformed.
    * The return value is a brand-new Promise, so it's valid to use this stage to transform a previously-rejected Promise into a fulfilled Promise.
  * Use [`$.when`](http://api.jquery.com/jQuery.when/) only with two or more argument to aggregate and gate multiple Promises.
    * Do not use `$.when` with zero arguments to create a new resolved Promise: it is non-intuitive to interpret. Use the more explicit `$.Deferred().resolve().promise()`.
    * Do not use `$.when` with one argument to re-wrap an existing Promise. Simply attach new handlers directly.
  * Use `$.Deferred()` to create new asynchronous results.
    * Only do this when it's not feasible to chain off an existing Promise with `.then`.
    * Scope the new Deferred within a closure, so it can be fulfilled with [`.resolve`](http://api.jquery.com/deferred.resolve/) or [`.reject`](http://api.jquery.com/deferred.reject/) eventually.
    * Do not return the new Deferred to an external scope directly, instead return [`.promise()`](http://api.jquery.com/promise/), as Promises are read-only.
    * [`.resolveWith`](http://api.jquery.com/deferred.resolveWith/) and [`.rejectWith`](http://api.jquery.com/deferred.rejectWith/) should typically not be necessary, as typical handlers will be bound with ES 6 arrow functions.
  * Don't use [`.notify`](http://api.jquery.com/deferred.notify/) or [`.progress`](http://api.jquery.com/deferred.progress/), as Girder doesn't yet have clear guidelines for them.
    * Eventually we should address whether this feature is worth using.

* [ ] All asynchronous operations in Girder, particularly Backbone Model / Collection methods, should return Promises.
    * Some of this was started in https://github.com/girder/girder/pull/1504
    * Many other methods still need to be migrated.
    * Where practical, some calls to `restRequest` inside Views should be moved inside the appropriate Model / Collection, so they're better encapsulated and more accessible.

* [ ] Girder functions returning Promises should be documented, to ensure it's part of their contract and not changed in the future.
    * [JSDoc does not have a comprehensive way to document Promise returns](https://github.com/jsdoc3/jsdoc/issues/1197)(https://github.com/jsdoc3/jsdoc/issues/1197), but something like `@returns {$.Promise} A promise, which fulfills with an X object, and rejects with a Y object.` should probably be adequate.
    * Where necessary or feasible, we should add unit tests to directly verify behavior and prevent regressions.
    * We should insist that newly added asynchronous methods adhere to these standards.

* [ ] Girder code should prefer consuming Promises to Backbone Events, where possible.
    * Event hookups should be reserved for long-term bindings (e.g. whenever a Model changes from any source, re-render the DOM that displays it).
    * If an asynchronous operation is being directly made and immediately listened for, Promises are much more robust.
    * For example:
      * [Code like this](https://github.com/girder/girder/blob/48b3661206919df699d578271558b74a1eff0c1e/clients/web/src/views/body/UserView.js#L131)
        ```
        this.model.off('g:saved').on('g:saved', function () {
            events.trigger('g:alert', ...);
            this.render();
        }, this).off('g:error').on('g:error', function (err) {
            events.trigger('g:alert', ...);
        }).save();
        ```
        is brittle and dangerous, and actually ends up having several bugs, side-effects, and race-conditions when any other code is also using `this.model`:
          * Any other listeners that may want to persistently bind to `'g:saved'` will be removed unconditionally.
          * Resolution does not unbind either of these listeners, so any other code that causes `'g:saved'` or `'g:error'` to trigger on the `this.model` at any point in the future will cause these handlers to execute again.
          * Using `.once` instead of `.on` will only unbind one event listener, not both.
          * Unbinding both event listeners within the handler will also unbind any external listeners too.
          * These particular listeners are not specifically bound to this particular `.save()` operation, so another parallel operation that triggers `'g:error'` on the model for any reason could cause this handler to execute (which would be even worse if the handler unbound the listener to `'g:saved'` within itself).

        This code is also harder to initially read, since all event handlers must be bound before the `.save()` occurs.
      * This can be refactored as
        ```
        this.model
            .save()
            .done(() => {
                events.trigger('g:alert', ...);
                this.render();
            })
            .fail(() => {
                events.trigger('g:alert', ...);
            })
        ```
        to be easier to comprehend, and much safer.
  * This will effectively resolve https://github.com/girder/girder/issues/1072, as `$.when` should be usable in the same circumstances.

**Comment by brianhelba** [comment]
@girder/developers Please excuse my wall of text, but these ideas have been percolating for a while. If the best practices guide for Promises actually makes sense to people, I may want to expand it with additional code samples and distribute it via a blog post, Girder docs, or something.

Most of the first issues are pretty easy to fix, and I'll be working on PRs for them. The later issues around moving away from Events and towards Promises will definitely require community consensus before they can be realized.

**Comment by kotfic** [comment]
Forgive me for not being more familiar with girder's use of promises,  but do you see a path forward that allows this over all program to happen in a non-breaking way?  Or is this something you see slated for girder 3.X?

**Comment by sebastienbarre** [comment]
You are already using webpack and babel, so might as well stick to more modern practices, e.g. async/await, or at least ES6 promises (via babel-polyfill or [es6-promise](https://github.com/stefanpenner/es6-promise) if needed, which provide a regenerator runtime). I'm not sure which part of Girder would have requirements in terms of Promise performance, but I might be wrong.

**Comment by brianhelba** [comment]
@kotfic I hope that nearly all of this can be done without a major API bump:
* This proposal already suggests a compatibility shim for `.error` for the the jQuery 2 -> 3 change.
* Most of the other breaking changes in jQuery 3 are minor enough to safely ignore, though we'll certainly need to look at specifics w.r.t. Girder. I believe that @jcfr has already done some work on this.
* Most of the methods that should return promises currently return nothing, so adding a return value is safe.
* There may be some places that return a non-promise value (that should), which we consider stable enough to be part of the API surface. Those would have to wait for Girder 3.
* The greater usage of promises within Girder would be primarily about refactoring implementations (and changing our standards for new code), and would not change the API surface.

**Comment by brianhelba** [comment]
@sebastienbarre At the moment, I think jQuery 3 promises are actually superior to ES6 promises. The speed issue is not so much about Girder performance, but more as evidence that ES6 features are not necessarily intrinsically superior to third-party library code. Since we will always have jQuery available, we should compare them on their feature merits. On that front, jQuery 3 Promise simply is a useful superset of ES6 Promise (with nice helper methods like `.done` and `fail` (`.tap` and `.tapCatch` in Bluebird)), though admittedly lacking ES6's `Promise.race`.

async/await requires a lot of compatibility code to run on a browser, and is a feature that relatively fewer people are familiar and comfortable with. I'd rather see progress towards improving Girder happen immediately (as the status quo is thoroughly backwards and broken, IMHO), so I'm trying to propose something that will be likely to have broad adoption by Girder developers and users (who may be at a variety of institutions, with a variety of Javascript familiarity). Given that Girder tries so hard to maintain semver compatibility, I'm advocating for the solution that requires the least number of radical changes while still bringing Girder to a place where asynchronous operations are handled in a sane, maintainable way.

**Comment by sebastienbarre** [comment]
My gut feeling is that when jQuery goes away from Girder -- and I believe it's bound to -- you will have less work to do if you adopt ES6 Promises now. It so happens you have them at the ready, whereas you do not have jQuery 3, correct? One of them is the way of the future (or the present, really). I would be nervous upgrading from jQuery 2 to 3, *especially* if the main motivation is to fix its broken Promise/deferred. The risk of borking the UI is not small -- I recall upgrading KWiK from jQuery 1.x to 2.x, that path was replete with tiny pitfalls. But maybe I missed another PR where upgrading to jQuery 3 was discussed. Isn't it a break in semver as well? I believe Girder exposes some jQuery widgets, are they guaranteed to be compatible?

**Comment by brianhelba** [comment]
The jQuery 2 -> 3 migration is *supposedly* easier than the 1 -> 2 one, but I readily admit that it's an unknown space for me, and could be more problematic than anticipated. Regardless, it needs to happen eventually. jQuery will go away no sooner than Backbone does.

Even if we consider ES6 promises to be the new baseline, we can still consider superior alternatives on their own merits (much like Underscore / Lodash is still useful even after ES5 got [`Array.prototype.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and [`Array.prototype.reduce`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)). IMHO (which others may legitimately disagree with), jQuery is slightly superior and Bluebird is significantly superior.

What I think is more relevant is the fact that Girder uses jQuery Promises right now, just rather poorly. Rather than rip it all up and change everything in an isolated `3.0-integration` branch, I'd rather start making some easy less-disruptive improvements right now. If a consensus can be built to wrap all `restResponse` in ES6 Promises or start using `async`/`await` in Girder 3+, I'd be onboard. Right now though, I'm too impatient to wait, as my own plugin development is necessitating that I rewrap or reimplement more and more of Girder's models to return promises. If Girder can be made to stop using Events in the wrong ways right now, that's a big enough improvement for me to want to move forward on.

**Comment by sebastienbarre** [comment]
> Regardless, it needs to happen eventually.

For some reason I thought Girder was going towards React at some point. You will not want to keep jQuery around when it happens. But maybe things have changed (all the shiny...).

**Comment by zachmullen** [comment]
We toyed with the idea of moving to react, but have not committed to it. At this point it's low on the priority list for the platform.
